Inital migration article port (#110)
dotnet/docs-desktop: dotnet-desktop-guide/net/winforms/migration/index.md

---
title: Migrate a Windows Forms app to .NET 5
description: Teaches you how to port a .NET Framework Windows Forms application to .NET 5.
ms.date: 11/02/2020
ms.topic: how-to
---

# How to migrate a Windows Forms desktop app to .NET 5

This article describes how to migrate your Windows Forms-based desktop app from .NET Framework to .NET 5 or later. The .NET SDK includes support for Windows Forms applications. Windows Forms is still a Windows-only framework and only runs on Windows.

Migrating your app from .NET Framework to .NET 5 generally requires a new project file. .NET 5 uses SDK-style project files while .NET Framework typically uses the older Visual Studio project file. If you've ever opened a Visual Studio project file in a text editor, you know how verbose it is. SDK-style projects are smaller and don't require as many entries as their counterparts do.

> To learn more about the benefits of .NET (not .NET Framework), see [Introduction to .NET](/dotnet/core/introduction).

Author: I don't know how I feel about this line...

To migrate the **packages.config** file, do the following:

01. In **Solution explorer**, find the project you're migrating.
02. Right-click on **packages.config** > **Migrate packages.config to ProjectReference**.
03. Select all of the top-level packages.

A build report is generated to let you know of any issues migrating the NuGet packages.

## Project file

The next step in migrating your app is converting the project file. As previously stated, .NET 5 uses SDK-style project files and won't load the Visual Studio project files that .NET Framework uses. However, there's the possibility that you're already using SDK-style projects. You can easily spot the difference in Visual Studio. Right-click on the project file in **Solution explorer** and look for the **Edit Project File** menu option. If this menu item is missing, you're using the old Visual Studio project format and need to upgrade.

To upgrade, do the following:

01. In Solution explorer,** find the project you're migrating.
01. Right-click on the project and select **Unload Project**.
01. Right-click on the project and select **Edit Project File**.

> 01. Copy-and-paste the project XML into a text editor. You'll want a copy so that it's easy move content into the new project.

Reviewer: ```suggestion
01. Copy-and-paste the project XML into a text editor. You'll want a copy so that it's easy to move content into the new project.
```

01. Erase the content of the file and paste in the following content:

    ```xml
    <Project Sdk="Microsoft.NET.Sdk">

      <PropertyGroup>
        <OutputType>WinExe</OutputType>
        <TargetFramework>net5.0-windows</TargetFramework>
        <UseWindowsForms>true</UseWindowsForms>
        <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
      </PropertyGroup>

    </Project>
    ```

    > [!IMPORTANT]
    > Libraries don't need to define an `<OutputType>` setting. Remove that entry if you're upgrading a library project.

> This XML gives you the basic structure of the project. However, it doesn't contain any of the settings from the old project file. Using the old project information you previously copied to a text editor, copy the following settings.

Reviewer: ```suggestion
This XML gives you the basic structure of the project. However, it doesn't contain any of the settings from the old project file. Using the old project information you previously copied to a text editor, do the following:
```

> 01. With the `<PropertyGroup>` element in the new project, copy the following from the old project:

Reviewer: ```suggestion
01. Copy the following elements from the old project file into the `<PropertyGroup>` element in the new project file: 
```

    - `<RootNamespace>`
    - `<AssemblyName>`

    Your project file should look similar to the following:

    ```xml
    <Project Sdk="Microsoft.NET.Sdk">

      <PropertyGroup>
        <OutputType>WinExe</OutputType>
        <TargetFramework>net5.0-windows</TargetFramework>
        <UseWindowsForms>true</UseWindowsForms>
        <GenerateAssemblyInfo>false</GenerateAssemblyInfo>

        <RootNamespace>MatchingGame</RootNamespace>
        <AssemblyName>MatchingGame</AssemblyName>
      </PropertyGroup>

    </Project>
    ```

> 01. After the `</PropertyGroup>` element closure, copy the `<ItemGroup>` entries from the old project that contain `<ProjectReference>` or `<PackageReference>`.

Reviewer: ```suggestion
01. Copy the `<ItemGroup>` elements from the old project file that contain `<ProjectReference>` or `<PackageReference>` into the new file after the `</PropertyGroup>` closing tag.
```

    Your project file should look similar to the following:

    ```xml
    <Project Sdk="Microsoft.NET.Sdk">

      <PropertyGroup>
        (contains settings previously described)
      </PropertyGroup>

      <ItemGroup>
        <ProjectReference Include="..\MatchingGame.Logic\MatchingGame.Logic.csproj">
          <Project>{36b3e6e2-a9ae-4924-89ae-7f0120ce08bd}</Project>
          <Name>MatchingGame.Logic</Name>
        </ProjectReference>
      </ItemGroup>
      <ItemGroup>
        <PackageReference Include="MetroFramework">
          <Version>1.2.0.3</Version>
        </PackageReference>
      </ItemGroup>

    </Project>
    ```

>     The `<ProjectReference>` elements don't need the `<Project>` and `<Name>` children, you can remove those:

Reviewer: ```suggestion
    The `<ProjectReference>` elements don't need the `<Project>` and `<Name>` children, so you can remove those:
```

    ```xml
    <ItemGroup>
      <ProjectReference Include="..\MatchingGame.Logic\MatchingGame.Logic.csproj" />
    </ItemGroup>
    ```

### Resources and settings

Windows Forms projects for .NET Framework typically include other files such as *Properties/Settings.settings* and *Properties/Resources.resx*. These files, and any *resx* file created for your app besides form *resx* files, would need to be migrated.

> Copy those entries from the old project file into an `<ItemGroup>` entry in the new project. After you copy the entries, change any `<Compile Include="value">` or `<EmbeddedResource Include="value">` to instead use `Update` instead of `Include`, otherwise you'll get a compile error.

Reviewer: ```suggestion
Copy those entries from the old project file into an `<ItemGroup>` element in the new project. After you copy the entries, change any `<Compile Include="value">` or `<EmbeddedResource Include="value">` elements to instead use `Update` instead of `Include`.
```

      <AutoGen>True</AutoGen>
      <DependentUpon>Settings.settings</DependentUpon>
      <DesignTimeSharedInput>True</DesignTimeSharedInput>
    </Compile>
  </ItemGroup>
  ```

- Import the configuration for any *resx* file, such as the *properties/Resources.resx* file. Note that `Include` was changed to `Update` on both the `<Compile>` and `<EmbeddedResource>` elements, and `<SubType>` was removed from `<EmbeddedResource>`:

  ```xml
  <ItemGroup>
    <EmbeddedResource Update="Properties\Resources.resx">
      <Generator>ResXFileCodeGenerator</Generator>
      <LastGenOutput>Resources.Designer.cs</LastGenOutput>
    </EmbeddedResource>
    <Compile Update="Properties\Resources.Designer.cs">
      <AutoGen>True</AutoGen>
      <DependentUpon>Resources.resx</DependentUpon>
      <DesignTime>True</DesignTime>
    </Compile>
  </ItemGroup>
  ```

Convert each project in your solution. If you're using the sample app previously referenced, the **MatchingGame.Logic** project would be converted.

## Edit App.config

If your app has an *App.config* file, remove the `<supportedRuntime>` element.

```xml
<supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.5" />
```

> There are some things you should consider with the *App.config* file. The *App.config* file in .NET Framework was used to not only configure the app, but to configure runtime settings and behavior, such as logging. The *App.config* file in .NET 5+ (and .NET Core) is no longer used for runtime configuration. If your *App.config* file has these sections, they either won't be respected.

Reviewer: ```suggestion
There are some things you should consider with the *App.config* file. The *App.config* file in .NET Framework was used not only to configure the app, but to configure runtime settings and behavior, such as logging. The *App.config* file in .NET 5+ (and .NET Core) is no longer used for runtime configuration. If your *App.config* file has these sections, they won't be respected.
```

## Add the compatibility package

> If you can't compile and receive errors similar to the following:

Reviewer: ```suggestion
If compilation fails and you receive errors similar to the following:
```

- **The type or namespace \<some name> could not be found**
- **The name \<some name> does not exist in the current context**

> You may need to add the [**Microsoft.Windows.Compatibility**](https://www.nuget.org/packages/Microsoft.Windows.Compatibility/) package to your app. This package adds ~21,000 .NET APIs from .NET Framework, such as the `System.Configuration.ConfigurationManager` class and interacting with the Windows Registry.

Reviewer: ```suggestion
You may need to add the [**Microsoft.Windows.Compatibility**](https://www.nuget.org/packages/Microsoft.Windows.Compatibility/) package to your app. This package adds ~21,000 .NET APIs from .NET Framework, such as the `System.Configuration.ConfigurationManager` class and APIs for interacting with the Windows Registry.
```

```xml
<ItemGroup>
  <PackageReference Include="Microsoft.Windows.Compatibility" Version="5.0.0-rc.2.20475.5" />
</ItemGroup>
```

## Test your app

> After you've finished migrating your app, test it! Testing is the best way to identify behavioral changes to your app or bugs you need to fix.

Reviewer: ```suggestion
After you've finished migrating your app, test it!
```
